File: Ash/Mode/6_pcmutex.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

#define BUFFER_SIZE 5

int buffer[BUFFER_SIZE];
int in = 0, out = 0;

pthread_mutex_t mutex_empty;
pthread_mutex_t mutex_filled;

void *producer(void *arg)
{
    int item = 1;
    while (1)
    {
        pthread_mutex_lock(&mutex_empty);

        buffer[in] = item;
        printf("Producer produced: %d\n", item);
        in = (in + 1) % BUFFER_SIZE;

        pthread_mutex_unlock(&mutex_filled);
        item++;
    }
    pthread_exit(NULL);
}

void *consumer(void *arg)
{
    int item;
    while (1)
    {

        pthread_mutex_lock(&mutex_filled);

        item = buffer[out];
        printf("Consumer consumed: %d\n", item);
        out = (out + 1) % BUFFER_SIZE;

        pthread_mutex_unlock(&mutex_empty);
    }
    pthread_exit(NULL);
}

int main()
{
    pthread_t producer_thread, consumer_thread;

    pthread_create(&producer_thread, NULL, producer, NULL);
    pthread_create(&consumer_thread, NULL, consumer, NULL);

    pthread_join(producer_thread, NULL);
    pthread_join(consumer_thread, NULL);

    return 0;
}

File: Ash/Mode/6_pcmutex.py
import threading

BUFFER_SIZE = 5
buffer = [0] * BUFFER_SIZE
in_index = 0
out_index = 0

mutex_empty = threading.Semaphore(BUFFER_SIZE)
mutex_filled = threading.Semaphore(0)

def producer():
    global in_index
    item = 1
    while True:
        mutex_empty.acquire()

        buffer[in_index] = item
        print(f"Producer produced: {item}")
        in_index = (in_index + 1) % BUFFER_SIZE

        mutex_filled.release()
        item += 1

def consumer():
    global out_index
    while True:
        mutex_filled.acquire()

        item = buffer[out_index]
        print(f"Consumer consumed: {item}")
        out_index = (out_index + 1) % BUFFER_SIZE

        mutex_empty.release()

if __name__ == "__main__":
    producer_thread = threading.Thread(target=producer)
    consumer_thread = threading.Thread(target=consumer)

    producer_thread.start()
    consumer_thread.start()

    producer_thread.join()
    consumer_thread.join()


#problem